### ISSUE TITLE
OperatorReplay leaks its subscribers

### ISSUE THREAD

**Issue by ntoskrnl** [ReactiveX/RxJava#4228]
We use `replay(1)` a lot in our Android projects. All subscriptions are kept in `CompositeSubscription` and, of course, we always clear it when we don't need it anymore.

Even though we unsubscribe from observable, subscriber won't get collected by GC, because it is referenced by `OperatorReplay`, which prevents Activity or Fragment from being garbage collected. We do need to keep observable alive and resubscribe to it from new Activity instance, but now we can't use replay() as all our Activities are leaked.

Right now we downgraded back to 1.1.5. Alternatively, one can use `publish()` with behavior subject to work around this issue. I believe the issue was introduced in one of the latest versions of the library (perhaps 1.1.6, not sure though).

**Comment by akarnokd** [comment]
Do you have your replay in a completed state or with infrequent updates?

**Comment by ntoskrnl** [comment]
Infinite stream is "replayed", so, I guess, it is never in a completed state. 
What do you mean by infrequent updates?

**Comment by ntoskrnl** [comment]
Connectable observable with replay example (no error handling to keep it simple): 

``` kotlin
val dataStream = refreshes.startWith(Unit)
        .flatMap { model.getData() }
        .replay(1)
```

When presenter is created we connect to the stream and store subscription until presenter is destroyed:

``` kotlin
val presenterSubscription = CompositeSubscription()
val viewSubscription = CompositeSubscription()

presenterSubscription.add(dataStream.connect())
```

View is attached to presenter:

``` kotlin
// subscribe view to the data stream, subscriber has a strong reference to view
viewSubscription.add(dataStream.subscribe {  view.showContent(it) })
```

View detached from presenter:

``` kotlin
// clear view subscription
// after this there should be no strong references to the view
viewSubscription.clear()
```

When presenter is destroyed (when activity is closed): 

``` kotlin
presenterSubscription.unsubscribe()
```

Throughout the presenter lifecycle, different instances of view are attached and then detached (for instance, during screen rotation when Activity is recreated). However, after the new view instance is attached, the old instance is still referenced from within OperatorReplay. According to heap dump analysis, each time there exist only two instances of the view – no matter how I many times I rotate the screen, there are two instances of view: current and only 1 leaked view.

**Comment by ntoskrnl** [comment]
👍

**Comment by akarnokd** [comment]
Thanks for the report. This is a bug in `replay()`'s internal cache of `Subscriber`s. I've posted a PR #4229 with the fix. A temporary workaround is to use `onTerminateDetach` on the `ConnectableObservable` the `replay()` returns and subscribe through it.

**Comment by akarnokd** [comment]
Fixed via #4229.
